### PR TITLE
Create WDL version of BCL2FASTQ

### PR DESCRIPTION
**Pull request by apmasell**

### CODE REVIEW

**Comment by apmasell** [comment]
retest this please

**Comment by apmasell** [comment]
retest this please

**Comment by apmasell** [comment]
retest this please

**Comment by apmasell** [comment]
retest this please

**Comment by apmasell** [comment]
retest this please

File: wrapper/main.cpp [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+            << " concatenation threads to finish." << std::endl;
+  for (auto &running_thread : running_threads) {
+    if (running_thread.joinable()) {
+      running_thread.join();
+    } else {
+      std::cerr << "Thread is not joinable. That shouldn't be possible."
+                << std::endl;
+    }
+  }
+  std::cerr << "Concenatentation finished." << std::endl;
```
Comment (avarsava) [inline wrapper/main.cpp:474]: Concatenation

File: wrapper/main.cpp [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+      }
+      if (inputFiles.empty()) {
+        // Write empty file
+        ogzstream(filename.str().c_str());
+        std::cerr << "No data for " << filename.str()
+                  << ". Creating empty file." << std::endl;
+      } else {
+        // Create a new thread for doing the concatentation.
+        std::cerr << "Starting thread for " << filename.str()
+                  << " conatentation from " << inputFiles.size() << " files."
```
Comment (avarsava) [inline wrapper/main.cpp:94]: concatenation

File: wrapper/main.cpp [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+}
+
+// Store information about a sample in the input (might map to multiple rows in
+// the sample sheet if multiple barcodes)
+class Sample {
+public:
+  Sample(Json::ArrayIndex id_, const std::string &name_, int barcodes_)
+      : id(id_), barcodes(barcodes_), name(name_) {}
+
+  // Find any matching files and spin up a concentation thread for each one;
```
[the remark this answers is not in the document]
Comment (apmasell) [reply wrapper/main.cpp:72]: I spelled this wrong in almost every place.

File: wrapper/main.cpp [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+                  << std::endl;
+        std::thread copier(concat_files, filename.str(), inputFiles);
+        threads.push_back(std::move(copier));
+      }
+    }
+    std::cerr << "Getting read count " << name << " from Stats/Stats.json."
+              << std::endl;
+
+    auto readCount = 0;
+    for (auto barcode = 0; barcode < barcodes; barcode++) {
```
Comment (asofmosq) [inline wrapper/main.cpp:129]: What's happening here??

File: wrapper/main.cpp [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+    for (Json::ArrayIndex j = 0; j < cr["Undetermined"]["ReadMetrics"].size();
+         j++) {
+      auto &rm = cr["Undetermined"]["ReadMetrics"][j];
+      for (Json::ArrayIndex k = 0; k < rm.size(); k++) {
+        reads.insert(rm["ReadNumber"].asInt());
+      }
+    }
+  }
+
+  std::vector<std::string> fastqs;
```
Comment (asofmosq) [inline wrapper/main.cpp:497]: Perhaps add one line here explaining that the fastq files are being extracted from the temporary directory?

**Comment by apmasell** [comment]
Updated.

File: bcl2fastq.wdl [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    String runDirectory
+    SampleList samples
+    String temporaryDirectory = "."
+    Int threads = 8
+    Int timeout = 40
+  }
+  parameter_meta {
+    basesMask: "An Illumina bases mask string to use. If absent, the one written by the instrument will be used."
+    bcl2fastq: "The name or path of the BCL2FASTQ executable."
+    bcl2fastqJail: "The name ro path of the BCL2FASTQ wrapper script executable."
```
[the remark this answers is not in the document]
Comment (avarsava) [reply bcl2fastq.wdl:89]: bump

File: bcl2fastq.wdl [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    extraOptions: "Any other options that will be passed directly to bcl2fastq."
+    ignoreMissingBcls: "Flag passed to bcl2fastq, allows missing bcl files."
+    ignoreMissingFilter: "Flag passed to bcl2fastq, allows missing or corrupt filter filesa."
```
[the remark this answers is not in the document]
Comment (avarsava) [reply bcl2fastq.wdl:92]: bump

**Pull request merged**